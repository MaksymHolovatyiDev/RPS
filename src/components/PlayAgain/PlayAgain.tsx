import {Animated, Text, TouchableOpacity} from 'react-native';
import {styles} from './PlayAgain.styles';
import {useRPS} from '@/Provider/Provider';
import {items, playerMessage} from '@/components/RPS/data';
import {useEffect, useRef, useState} from 'react';
import {AnimatedFunction} from 'src/helpers/helpers';

export default function PlayAgain() {
  const [message, setMessage] = useState('');
  const {picked, housePick, setPicked, setScore, opacity} = useRPS();
  const translateY = useRef(new Animated.Value(60)).current;
  const mainOpacity = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    setTimeout(() => {
      const playerIdx = items.findIndex(el => picked?.key === el.key);
      const resultMsg = playerMessage[playerIdx][housePick!];
      setMessage(resultMsg);
      if (resultMsg === 'You win') setScore(prevState => ++prevState);
      AnimatedFunction(translateY, 0, 300);
      AnimatedFunction(mainOpacity, 1, 300);
    }, 1200);
  }, []);

  return (
    <Animated.View
      style={{
        ...styles.container,
        opacity: mainOpacity,
        transform: [{translateY}],
      }}>
      <Text style={styles.text}>{message}</Text>
      <TouchableOpacity
        style={styles.btn}
        activeOpacity={0.8}
        onPress={() => {
          setPicked(null);
          AnimatedFunction(opacity, 1);
        }}>
        <Text style={styles.btnText}>Play again</Text>
      </TouchableOpacity>
    </Animated.View>
  );
}
