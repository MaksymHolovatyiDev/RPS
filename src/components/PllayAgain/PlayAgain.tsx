import { Text, TouchableOpacity, View } from 'react-native';
import { styles } from './PlayAgain.styles';
import { useRPS } from '@/Provider/Provider';
import { items } from '@/components/RPS/data';
import { useEffect, useState } from 'react';

export default function PlayAgain() {
  const [message, setMessage] = useState('');
  const { picked, housePick, score, setPicked, setScore } = useRPS();

  useEffect(() => {
    if (picked?.key === items[housePick ? housePick - 1 : 2].key) {
      setScore(prevState => ++prevState);
      setMessage('You win');
    } else if (picked?.key === items[housePick!].key) {
      setMessage('Draw');
    } else {
      setMessage('You lose');
    }
  }, []);

  return (
    <View style={styles.container}>
      <Text style={styles.text}>{message}</Text>
      <TouchableOpacity
        style={styles.btn}
        activeOpacity={0.8}
        onPress={() => setPicked(null)}
      >
        <Text style={styles.btnText}>Play again</Text>
      </TouchableOpacity>
    </View>
  );
}
