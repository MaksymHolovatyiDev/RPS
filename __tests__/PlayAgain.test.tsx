import {render, screen, waitFor} from '@testing-library/react-native';

import PlayAgain from 'src/components/PlayAgain/PlayAgain';

let mockHousePick = 1;

jest.mock('src/Provider/Provider', () => {
  return {
    useRPS: () => ({
      picked: {key: 'Scissors'},
      housePick: mockHousePick,
      setScore: () => {},
    }),
  };
});
jest.mock('src/helpers/helpers', () => ({
  AnimatedFunction: () => {},
}));

describe('Play again render', () => {
  it('Should renders correctly', () => {
    const tree = render(<PlayAgain />).toJSON();

    expect(screen.getByText('Play again')).toBeDefined();
    expect(tree).toMatchSnapshot();
  });

  it('Should win', async () => {
    render(<PlayAgain />);

    await waitFor(
      () => {
        expect(screen.getByText('You win')).toBeDefined();
      },
      {timeout: 1300},
    );
  });

  it('Should lose', async () => {
    mockHousePick = 4;

    render(<PlayAgain />);

    await waitFor(
      () => {
        expect(screen.getByText('You lose')).toBeDefined();
      },
      {timeout: 1300},
    );
  });

  it('Should to be draw', async () => {
    mockHousePick = 0;

    render(<PlayAgain />);

    await waitFor(
      () => {
        expect(screen.getByText('Draw')).toBeDefined();
      },
      {timeout: 1300},
    );
  });
});
