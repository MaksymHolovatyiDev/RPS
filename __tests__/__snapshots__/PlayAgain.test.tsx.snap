// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Play again render Should renders correctly 1`] = `
<View
  collapsable={false}
  style={
    {
      "alignItems": "center",
      "gap": 10,
      "justifyContent": "center",
      "opacity": 0,
      "position": "relative",
      "transform": [
        {
          "translateY": 60,
        },
      ],
    }
  }
>
  <Text
    style={
      {
        "color": "#fff",
        "fontFamily": "BarlowSemiCondensed-Bold",
        "fontSize": 48,
        "letterSpacing": 2,
        "textTransform": "uppercase",
      }
    }
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 8,
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 40,
        "paddingVertical": 14,
        "width": 240,
      }
    }
  >
    <Text
      style={
        {
          "color": "hsl(229, 25%, 31%)",
          "fontFamily": "BarlowSemiCondensed-SemiBold",
          "fontSize": 16,
          "letterSpacing": 2,
          "textTransform": "uppercase",
        }
      }
    >
      Play again
    </Text>
  </View>
</View>
`;
